applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_relaxed_gradient_projection.py

```diff
@@ -0,0 +1,576 @@
+                if constraint["type"].GetString() == "=":
+                    self.constraint_buffer_variables[identifier]["buffer_value"] = min(1 - abs(constraint_value) / self.constraint_buffer_variables[identifier]["buffer_size"], 2.0)
+                else:
+                    self.constraint_buffer_variables[identifier]["buffer_value"] = min( (constraint_value - self.constraint_buffer_variables[identifier]["lower_buffer_value"]) / self.constraint_buffer_variables[identifier]["buffer_size"], 2.0 )
+            else:
+                self.constraint_buffer_variables[identifier]["buffer_value"] = 0.0
+
+    # --------------------------------------------------------------------------
+    def __computeShapeUpdate(self):
+        self.mapper.Update()
+        self.mapper.InverseMap(KSO.DF1DX, KSO.DF1DX_MAPPED)
+
+        for constraint in self.constraints:
+            con_id = constraint["identifier"].GetString()
+            gradient_variable = self.constraint_gradient_variables[con_id]["gradient"]
+            mapped_gradient_variable = self.constraint_gradient_variables[con_id]["mapped_gradient"]
+            self.mapper.InverseMap(gradient_variable, mapped_gradient_variable)
+
+        self.inner_iter = 1
+
+        #TODO add convergence criteria, for now only 1 iteration run for testing
+        while not self.__checkInnerConvergence():
+
+            self.direction_has_changed = False
+
+            KM.Logger.PrintInfo("ShapeOpt", "Inner Iteration to Find Shape Update = ", self.inner_iter)
+
+            self.__computeControlPointUpdate()
+
+            self.mapper.Map(KSO.CONTROL_POINT_UPDATE, KSO.SHAPE_UPDATE)
+            self.model_part_controller.DampNodalUpdateVariableIfSpecified(KSO.SHAPE_UPDATE)
+
+            self.__checkConstraintValue()
+            self.inner_iter += 1
+        self.__saveLineSearchData()
+
+
+    # --------------------------------------------------------------------------
+
+    def __checkInnerConvergence(self):
+        KM.Logger.PrintInfo("Check Convergence of the inner loop:")
+        if self.inner_iter == 1:
+            return False
+        elif self.direction_has_changed and self.inner_iter < 100:
+            return False
+        else:
+            return True
+
+    def __checkConstraintValue(self):
+        index = -1
+        for constraint in self.constraints:
+            if self.__isConstraintActive(constraint):
+                index += 1
+                identifier = constraint["identifier"].GetString()
+                constraint_value = self.communicator.getStandardizedValue(identifier)
+                g_a_variable = self.constraint_gradient_variables[identifier]["gradient"]
+                shape_update = KM.Vector()
+                gradient = KM.Vector()
+                self.optimization_utilities.AssembleVector(self.design_surface, gradient, g_a_variable)
+                self.optimization_utilities.AssembleVector(self.design_surface, shape_update, KSO.SHAPE_UPDATE)
```

Comment:
This is always the same, you could assemble it once outside the loop. Or if you do not want to do that, initialize it outside the loop with 
```
shape_update = None
```
and in the loop 
```
if not shape_update:
                     self.optimization_utilities.AssembleVector(self.design_surface, shape_update, KSO.SHAPE_UPDATE)
```